bin/generate

```diff
@@ -1,38 +1,65 @@
 #!/usr/bin/env python
 
+import argparse
+import enum
 import json
 import re
 import shutil
 
 from pathlib import Path
 from collections import OrderedDict
-from urllib.request import urlopen, HTTPError
+from urllib.request import urlopen, URLError
 from http import HTTPStatus
 
 
+@enum.unique
+class ProblemSpecsSource(enum.Enum):
+    Local = 'local'
+    Remote = 'remote'
+
+
 canonical_data_url = 'https://raw.githubusercontent.com/exercism' + \
                      '/problem-specifications/master' + \
                      '/exercises/%s/canonical-data.json'
 
+def smldir():
+    """Absolute path of the sml directory."""
+    return Path(__file__).absolute().parent.parent
+
+
+class CanonicalDataNotFound(Exception):
+    pass
 
-def fetch(exercise):
+
+def get_canonical_data_from_remote(exercise):
     try:
         with urlopen(canonical_data_url % exercise) as response:
             return json.loads(response.read().decode('utf8'))
-    except HTTPError as e:
-        if e.status == HTTPStatus.NOT_FOUND.value:
-            msg = "canonical-data.json was not found."
-        else:
-            msg = 'Unexpected error, please try again later.'
-        raise Exception(msg)
+    except URLError as e:
+        raise CanonicalDataNotFound(f"{e.reason}")
 
 
-def load(exercise):
-    root = Path(__file__).absolute().parent.parent.parent
-    path = root / 'problem-specifications/exercises'
-    with (path / exercise / 'canonical-data.json').open() as f:
-        return json.loads(f.read())
+def get_canonical_data_from_local(exercise):
+    path = smldir().parent / 'problem-specifications' / 'exercises'
 
+    try:
+        with (path / exercise / 'canonical-data.json').open() as f:
+            return json.loads(f.read())
+    except OSError as e:
+        raise CanonicalDataNotFound(f"{e}")
+
+
+def get_canonical_data(exercise, problem_specs_source=ProblemSpecsSource.Remote):
+    """Get canonical-data of the exercise as a json string."""
+    if problem_specs_source == ProblemSpecsSource.Remote:
+        return get_canonical_data_from_remote(exercise)
+    elif problem_specs_source == ProblemSpecsSource.Local:
+        return get_canonical_data_from_local(exercise)
+
+    raise Exception(
+        f"Unknown source for problem-specifications: '{problem_specs_source}'."
+        f" Valid options are {list(ProblemSpecsSource)}."
```

Comment:
Nice.

Reply:
@kotp I observed too, that after applying `generate` to some exercise the exercise may be broken (i.e. `make test-<exercise>` fails). But this is already an older issue (unrelated to this PR).

Mostly just the name of the function to be tested is wrong/different (e.g. `two-fer` vs `name`, but this is not the only example). If I recall correctly the generate-script assumes that the `property` in canonical-data is the name of the function. It seems that this changed some time ago for some exercises. However, the file `.meta/example.sml` still uses the old name. For some reason in two-fer the function was called `name` previously and this is also what is used at the web-site (when someone tries the exercise). This seems to cause the error. And as I said, other exercises are affected too.